src/util/base.cc

```diff
@@ -0,0 +1,207 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,          //
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF    //
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.               //
+ */
+
+#include "util/base.h"
+#include <assert.h>
+#include <memory>
+#include "core/util/base64.h"
+#include "core/util/exception.h"
+#include "core/util/log.h"
+
+namespace bpo = boost::program_options;
+
+static_assert(
+    Base32Command::m_BufferSize % 40 == 0,
+    "Invalid BASE32 buffer size");
+static_assert(
+    Base64Command::m_BufferSize % 12 == 0,
+    "Invalid BASE64 buffer size");
+
+BaseCommand::BaseCommand() : m_Desc("General options")
+{
+  m_Desc.add_options()("help,h", "produce this help message")(
+      "type,t", bpo::value<std::string>(&m_OptType), "encode/decode")(
+      "infile,i", bpo::value<std::string>(&m_OptInfile), "input file")(
+      "outfile,o", bpo::value<std::string>(&m_OptOutfile), "output file");
+}
+
+void BaseCommand::PrintUsage(const std::string& name) const
+{
+  LOG(info) << "Syntax: " << name << " encode <inFile> <outfile>";
+  LOG(info) << "or    : " << name << " decode <inFile> <outfile>";
+}
+
+template <typename ToBase, typename ToByte, typename SizeConst>
+bool process(
+    ToBase ByteStreamToBase,
+    ToByte BaseToByteStream,
+    SizeConst Size,
+    bool encode,
+    kovri::core::InputFileStream* input,
+    kovri::core::OutputFileStream* output)
+{
+  std::uint8_t in_buffer[Size + 1];
+  std::uint8_t out_buffer[Size * 2 + 1];
+  std::size_t ret(0);
+  memset(in_buffer, 0, sizeof(in_buffer));
+  memset(out_buffer, 0, sizeof(out_buffer));
+  do
+    {
+      if (!input->Read(in_buffer, Size))
+        return false;
+
+      if (encode)
+        ret = ByteStreamToBase(
+            in_buffer,
+            input->Count(),
+            reinterpret_cast<char*>(out_buffer),
+            Size * 2);
+      else
+        ret = BaseToByteStream(
+            reinterpret_cast<char*>(in_buffer),
+            input->Count(),
+            out_buffer,
+            Size * 2);
+
+      if (input->Count() && !ret)
+        {
+          LOG(error) << "Error : Stream processing failed !";
+          return false;
+        }
+
+      if (!output->Write(out_buffer, ret))
+        return false;
+    } while (input->Good());
+  return true;
+}
+
+bool BaseCommand::Impl(
+    const std::string& cmd_name,
+    int argc,
+    const char* argv[])
+{
+  bpo::variables_map vm;
+  bpo::positional_options_description pos;
+  pos.add("type", 1).add("infile", 1).add("outfile", 1);
+  try
+    {
+      bpo::parsed_options parsed = bpo::command_line_parser(argc, argv)
+                                       .options(m_Desc)
+                                       .positional(pos)
+                                       .run();
+      bpo::store(parsed, vm);
+      bpo::notify(vm);
+    }
+  catch (...)
```

Comment:
@guzzijones as I implied in https://github.com/monero-project/kovri/pull/556#discussion_r100957162, we've been using an exception dispatcher since #544. See `src/core/util/exception.{cc,h}`

@guzzijones @MoroccanMalinois the idea *is* to catch all exceptions, dispatch, log, and return control; allowing the freedom to re-throw if needed. @guzzijones https://isocpp.org/